README
ibus-hangul
-----------

ibus-hangul is a Korean input method engine for IBus. It is 

The official web site is:

    http://code.google.com/p/ibus/


Installation
------------

ibus-hangul requires ibus.

Install procedure:
 $ ./configure
 $ make
 $ make install


Bug report
----------

http://code.google.com/p/ibus/issues/list


NEWS
1.4.0
 * Require ibus 1.4
 * Add desktop menu for ibus-setup-hangul
 * Add an option for automatic reordering
 * Add an option for a word preedit text
 * Update MS IME compatible symbol table
 * Use surrounding text feature on converting to hanja


src/engine.h
/* vim:set et sts=4: */
/* ibus-hangul - The Hangul Engine For IBus
 * Copyright (C) 2008-2009 Peng Huang <shawn.p.huang@gmail.com>
 * Copyright (C) 2009-2011 Choe Hwanjin <choe.hwanjin@gmail.com>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 */

#ifndef __ENGINE_H__
#define __ENGINE_H__

#include <ibus.h>

#define IBUS_TYPE_HANGUL_ENGINE	\
	(ibus_hangul_engine_get_type ())

GType   ibus_hangul_engine_get_type    (void);

void    ibus_hangul_init (IBusBus *bus);
void    ibus_hangul_exit (void);

#endif


setup/main.py
# vim:set et sts=4 sw=4:
#
# ibus-hangul - The Hangul Engine For IBus
#
# Copyright (c) 2009-2011 Choe Hwanjin <choe.hwanjin@gmail.com>
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License along
# with this program; if not, write to the Free Software Foundation, Inc.,
# 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.

import sys
import os
import gobject
import gtk
import ibus
import locale
import gettext
import config
import subprocess
from keycapturedialog import KeyCaptureDialog

_ = lambda a : gettext.dgettext(config.gettext_package, a)

class Setup ():
    def __init__ (self):
        self.__bus = ibus.Bus()
        self.__config = self.__bus.get_config()
	self.__config.connect("value-changed", self.on_value_changed, None)

	ui_file = os.path.join(os.path.dirname(__file__), "setup.ui")
	self.__builder = gtk.Builder()
	self.__builder.set_translation_domain(config.gettext_package)
	self.__builder.add_from_file(ui_file)

	# Hangul tab
	pipe = subprocess.Popen([config.setuplibdir + '/hangul_keyboard_list'], stdout = subprocess.PIPE)
	list = pipe.communicate()[0].split('\n')
	
	self.__hangul_keyboard = self.__builder.get_object("HangulKeyboard")
	model = gtk.ListStore(str, str, int)
	i = 0
	for line in list:
	    items = line.split('\t')
	    if len(items) > 1:
		model.append([items[1], items[0], i])
		i+=1

	self.__hangul_keyboard.set_model(model)
	renderer = gtk.CellRendererText()
	self.__hangul_keyboard.pack_start(renderer)
	self.__hangul_keyboard.add_attribute(renderer, "text", 0)

	current = self.__read("HangulKeyboard", "2")
	for i in model:
	    if i[1] == current:
		self.__hangul_keyboard.set_active(i[2])
		break

	self.__word_commit = self.__builder.get_object("WordCommit")

	word_commit = self.__read("WordCommit", False)
        self.__word_commit.set_active(word_commit)

	self.__auto_reorder = self.__builder.get_object("AutoReorder")
	auto_reorder = self.__read("AutoReorder", True)
        self.__auto_reorder.set_active(auto_reorder)

	# hanja tab
	button = self.__builder.get_object("HanjaKeyListAddButton")
	button.connect("clicked", self.on_hanja_key_add, None)

	button = self.__builder.get_object("HanjaKeyListRemoveButton")
	button.connect("clicked", self.on_hanja_key_remove, None)

	model = gtk.ListStore(str)

	keylist_str = self.__read("HanjaKeys", "Hangul_Hanja,F9")
	self.__hanja_key_list_str = keylist_str.split(',')
	for i in self.__hanja_key_list_str:
	    model.append([i])

	self.__hanja_key_list = self.__builder.get_object("HanjaKeyList")
	self.__hanja_key_list.set_model(model)
	column = gtk.TreeViewColumn()
	column.set_title("key")
	renderer = gtk.CellRendererText()
	column.pack_start(renderer)
	column.add_attribute(renderer, "text", 0)
	self.__hanja_key_list.append_column(column)


	# advanced tab
	notebook = self.__builder.get_object("SetupNotebook")
	notebook.remove_page(2)

	# setup dialog
	self.__window = self.__builder.get_object("SetupDialog")
	icon_file = os.path.join(config.pkgdatadir, "icons", "ibus-hangul.svg")
	self.__window.set_icon_from_file(icon_file)
	self.__window.connect("response", self.on_response, None)
	self.__window.show()

	ok_button = self.__builder.get_object("button_cancel")
	ok_button.grab_focus()

    def run(self):
	res = self.__window.run()
	if (res == gtk.RESPONSE_OK):
	    self.on_ok()
	self.__window.destroy()

    def apply(self):
	model = self.__hangul_keyboard.get_model()
	i = self.__hangul_keyboard.get_active()
	self.__write("HangulKeyboard", model[i][1])

        word_commit = self.__word_commit.get_active()
        self.__write("WordCommit", word_commit)

	model = self.__hanja_key_list.get_model()
	str = ""
	iter = model.get_iter_first()
	while iter:
	    if len(str) > 0:
		str += ","
		str += model.get_value(iter, 0)
	    else:
		str += model.get_value(iter, 0)
	    iter = model.iter_next(iter)
	self.__write("HanjaKeys", str)

    def on_response(self, widget, id, data = None):
	if id == gtk.RESPONSE_APPLY:
	    self.apply()
	    widget.emit_stop_by_name("response")
	if id == gtk.RESPONSE_NONE:
	    widget.emit_stop_by_name("response")

    def on_ok(self):
	self.apply()

    def on_hanja_key_add(self, widget, data = None):
	dialog = KeyCaptureDialog(_("Select Hanja key"), self.__window)
	res = dialog.run()
	if res == gtk.RESPONSE_OK:
	    key_str = dialog.get_key_string()
	    if len(key_str) > 0:
		model = self.__hanja_key_list.get_model()
		iter = model.get_iter_first()
		while iter:
		    str = model.get_value(iter, 0)
		    if str == key_str:
			model.remove(iter)
			break
		    iter = model.iter_next(iter)

		model.append([key_str])
	dialog.destroy()

    def on_hanja_key_remove(self, widget, data = None):
	selection = self.__hanja_key_list.get_selection()
	(model, iter) = selection.get_selected()
	if model and iter:
	    model.remove(iter)

    def on_value_changed(self, config, section, name, value, data):
	if section == "engine/Hangul":
	    if name == "HangulKeyboard":
		model = self.__hangul_keyboard.get_model()
		for i in model:
		    if i[1] == value:
			self.__hangul_keyboard.set_active(i[2])
			break
	    elif name == "HanjaKeys":
		self.__hanja_key_list_str = value.split(',')

    def __read(self, name, v):
        return self.__config.get_value("engine/Hangul", name, v)

    def __write(self, name, v):
        return self.__config.set_value("engine/Hangul", name, v)

if __name__ == "__main__":
    locale.bindtextdomain(config.gettext_package, config.localedir)
    locale.bind_textdomain_codeset(config.gettext_package, "UTF-8")

    Setup().run()


setup/config.py
# vim:set et sts=4 sw=4:
#
# ibus-hangul - The Hangul Engine For IBus
#
# Copyright (c) 2009-2011 Choe Hwanjin <choe.hwanjin@gmail.com>
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License along
# with this program; if not, write to the Free Software Foundation, Inc.,
# 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.

gettext_package = 'ibus-hangul'
localedir = '/home/krisna/share/locale'
pkgdatadir = '/home/krisna/share/ibus-hangul'
pkglibdir = '/home/krisna/lib/ibus-hangul'
setupdatadir = pkgdatadir + '/setup'
setuplibdir = pkglibdir + '/setup'


setup/hangul_keyboard_list.c
/* vim:set et sw=4 sts=4: */
/* ibus-hangul - The Hangul Engine For IBus
 * Copyright (C) 2010-2011 Choe Hwanjin <choe.hwanjin@gmail.com>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 */

#include <stdio.h>
#include <locale.h>
#include <hangul.h>

int main()
{
    int n;
    int i;

    setlocale(LC_ALL, "");

    n = hangul_ic_get_n_keyboards();
    for (i = 0; i < n; ++i) {
	const char* id;
	const char* name;

	id = hangul_ic_get_keyboard_id(i);
	name = hangul_ic_get_keyboard_name(i);

	printf("%s\t%s\n", id, name);
    }
    
    return 0;
}


src/main.c
/* vim:set et sts=4: */
/* ibus-hangul - The Hangul Engine For IBus
 * Copyright (C) 2008-2009 Peng Huang <shawn.p.huang@gmail.com>
 * Copyright (C) 2009-2011 Choe Hwanjin <choe.hwanjin@gmail.com>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 */

#ifdef HAVE_CONFIG_H
#include <config.h>
#endif

#include <ibus.h>
#include <stdlib.h>
#include <locale.h>
#include <hangul.h>

#include "i18n.h"
#include "engine.h"


static IBusBus *bus = NULL;
static IBusFactory *factory = NULL;

/* options */
static gboolean ibus = FALSE;
static gboolean verbose = FALSE;

static const GOptionEntry entries[] =
{
    { "ibus", 'i', 0, G_OPTION_ARG_NONE, &ibus, "component is executed by ibus", NULL },
    { "verbose", 'v', 0, G_OPTION_ARG_NONE, &verbose, "verbose", NULL },
    { NULL },
};


static void
ibus_disconnected_cb (IBusBus  *bus,
                      gpointer  user_data)
{
    g_debug ("bus disconnected");
    ibus_quit ();
}


static void
start_component (void)
{
    IBusComponent *component;

    ibus_init ();

    bus = ibus_bus_new ();
    g_signal_connect (bus, "disconnected", G_CALLBACK (ibus_disconnected_cb), NULL);

    ibus_hangul_init (bus);

    component = ibus_component_new ("org.freedesktop.IBus.Hangul",
                                    N_("Korean input method"),
                                    "0.1.0",
                                    "GPL",
                                    "Peng Huang <shawn.p.huang@gmail.com>",
                                    "http://code.google.com/p/ibus/",
                                    "",
                                    "ibus-hangul");
    ibus_component_add_engine (component,
                               ibus_engine_desc_new ("hangul",
                                                     N_("Korean Input Method"),
                                                     N_("Korean Input Method"),
                                                     "ko",
                                                     "GPL",
                                                     "Peng Huang <shawn.p.huang@gmail.com>",
                                                     PKGDATADIR"/icon/ibus-hangul.svg",
                                                     "us"));

    factory = ibus_factory_new (ibus_bus_get_connection (bus));

    ibus_factory_add_engine (factory, "hangul", IBUS_TYPE_HANGUL_ENGINE);

    if (ibus) {
        ibus_bus_request_name (bus, "org.freedesktop.IBus.Hangul", 0);
    }
    else {
        ibus_bus_register_component (bus, component);
    }

    g_object_unref (component);

    ibus_main ();

    ibus_hangul_exit ();
}

int
main (gint argc, gchar **argv)
{
    GError *error = NULL;
    GOptionContext *context;

    setlocale (LC_ALL, "");

    bindtextdomain(GETTEXT_PACKAGE, LOCALEDIR);
    bind_textdomain_codeset(GETTEXT_PACKAGE, "UTF-8");
    textdomain(GETTEXT_PACKAGE);

    context = g_option_context_new ("- ibus hangul engine component");

    g_option_context_add_main_entries (context, entries, "ibus-hangul");

    if (!g_option_context_parse (context, &argc, &argv, &error)) {
        g_print ("Option parsing failed: %s\n", error->message);
        exit (-1);
    }

    start_component ();
    return 0;
}
